vassal-app/src/main/java/VASSAL/build/module/map/CounterDetailViewer.java

```diff
@@ -768,7 +769,20 @@ public Object visitStack(Stack s) {
       final boolean addContents = foundPieceAt == null ?
         super.visitStack(s) != null : foundPieceAt.equals(s.getPosition());
       if (addContents) {
-        s.asList().forEach(this::apply);
+        if (!(map instanceof PlayerHand)) {
+          s.asList().forEach(this::apply);
+        }
+        else {
+          for (final GamePiece p : s.asList()) {
+            final int index = s.indexOf(p);
+            if (index >= 0 && index < shapes.length) {
```

Comment:
How can `index` ever be negative or off the end here? `p` has to be in the list, since you got it from iterating over the list.

Reply:
I copied this out of PieceFinder.MOVABLE which had some scary comments about "blah blah blah THREADS, fear fear fear" and then did it that way.

Reply:
Hmm. Alright, I'll go look at that.

Reply:
I think you could do this:
```
for (int i = 0; i < shapes.length; ++i) {
  if (shapes[i].contains(pt)) {
    apply(s.getPieceAt(i));
    break;
  }
}
```